Enable calling get_runtime_delegate from app context

rseanhall opened the pull request:
This is the first part of implementing https://github.com/dotnet/runtime/issues/35465. This pull request is supposed to address all the complexities of allowing `hostfxr_get_runtime_delegate` to be called from a context from `hostfxr_initialize_for_dotnet_command_line`. As designed in https://github.com/dotnet/runtime/pull/36990, an app context may only ask for `hdt_load_assembly_and_get_function_pointer`. Implementing the new `hdt_get_function_pointer` type will be done in a subsequent pull request after this one is merged.

As part of this change, there are a couple of behavior changes that may technically be breaking changes.

### Loading the runtime when given an invalid delegate type

In 3.x, `hostfxr_get_runtime_delegate` doesn't error out when given an invalid delegate type. This means that it will load the runtime before calling into `hostpolicy`. Then the `get_delegate` call in `hostpolicy` is where it errors out. This behavior is unacceptable because the native host needs to have control over starting the runtime if there's no chance of getting a delegate from `hostfxr`.

In order to implement this change, I started making `hostpolicy` return the highest delegate type it knows about. Then `hostfxr` will check against that (or `hdt_load_assembly_and_get_function_pointer` if it's an old `hostpolicy`) before trying to load the runtime.

### ~Error code when trying to load SCD component~

~In 3.x, `hostfxr_initialize_for_runtime_config` returns the error code `InvalidConfigFile` when the config is for an SCD component. This error code is misleading since the config file is valid, just not currently supported by `hostfxr`. I made `hostfxr` return the new `HostApiUnsupportedScenario` error code in this case to give the native host better error information.~

rseanhall committed:
Add ability for hostfxr_get_runtime_delegate to block on invalid delegate types.

This is required so that the runtime isn't loaded if the request has no chance of succeeding.

Fix spelling of initialization_options_t.

Add context_contract_version_set flag to initialization_options_t.

This allows hostpolicy to know whether it can rely on the version field of corehost_context_contract to be valid.
Also always initialize corehost_context_contract to {}.

rseanhall committed:
Allow app-based context to call hostfxr_get_runtime_delegate.

Currently, it may only request the load_assembly_and_get_function_pointer delegate.

Fix discrepancy in design doc.

Add error code HostApiUnsupportedScenario for blocks that might be removed. This gives the native host better information on why the request failed.

Add tests for running a component from an app context.

vitek-karas commented:
Can you please fix the design doc in this PR as well?. In this [section](https://github.com/dotnet/runtime/blob/master/docs/design/features/native-hosting.md#getting-a-delegate-for-runtime-functionality) it states that the `hdt_load_assembly_and_get_function_pointer` is not allowed on "app contexts" - but later on it says it is allowed. I think it's fine to allow it, we just need to fix the doc.

vitek-karas commented:
Re breaking changes:
> Loading the runtime when given an invalid delegate type

I think this is good - you're right that we should not do any work if we know it's going to fail. So I like this change in general (I have yet to look at the implementation details).
As for the risk of taking this - I think it's very small, we still fail, and we avoid loading runtime. This is very unlikely to break people (not counting that it's very unlikely that anybody released an app which calls the API with the wrong value as it would not work anyway).

> Error code when trying to load SCD component

I'm not sure about this one. I do agree that the returned value is not great (misleading to say the least). That said I know about several cases where the exact exist code is used by calling code to make decisions. So changing it feels risky to me. If this would be a change in the runtime I would not worry too much as we're OK taking small breaking changes in major releases (with good reason). Unfortunately this is a change in `hostfxr` which doesn't really have the same freedom - it's global and callers always end up calling the latest version. So even minor breaking changes are a problem.

I think we should not make this change - I know the number value is bad, but we at least print out the right error message.

vitek-karas commented on src/installer/corehost/cli/corehost_context_contract.h:
```diff
@@ -25,7 +25,8 @@ enum class coreclr_delegate_type
     com_unregister,
-    load_assembly_and_get_function_pointer
+    load_assembly_and_get_function_pointer,
+    unknown, // Sentinel value for determining the last known delegate type
```
I would call it `__last` or something like that - make it stand out as "internal" value (double underscores) and "unknown" doesn't sound right to me, but I guess `__unknown` would be OK.

rseanhall replied:
I see now that `__last` is the convention in this area of the code so I'll go with that.

vitek-karas commented on src/installer/corehost/cli/corehost_context_contract.h:
```diff
@@ -38,6 +39,7 @@ struct corehost_initialize_request_t
 static_assert(offsetof(corehost_initialize_request_t, version) == 0 * sizeof(size_t), "Struct offset breaks backwards compatibility");
 static_assert(offsetof(corehost_initialize_request_t, config_keys) == 1 * sizeof(size_t), "Struct offset breaks backwards compatibility");
 static_assert(offsetof(corehost_initialize_request_t, config_values) == 3 * sizeof(size_t), "Struct offset breaks backwards compatibility");
+static_assert(sizeof(corehost_initialize_request_t) == 5 * sizeof(size_t), "Did you add static asserts for the newly added fields?");
```
Nice: thanks for adding this.

vitek-karas commented on src/installer/corehost/cli/corehost_context_contract.h:
```diff
@@ -59,6 +61,7 @@ struct corehost_context_contract
     int (HOSTPOLICY_CALLTYPE *get_runtime_delegate)(
         coreclr_delegate_type type,
         /*out*/ void **delegate);
+    size_t unknown_delegate_type; // Added in 5.0
```
If we return the first unknown I would call this `first_unknown_delegate_type`, or maybe `first_unsupported_delegate_type`.

Alternatively we could change the meaning to the last known/supported `last_supported_delegate_type`.

rseanhall replied:
I think I'm going to change the meaning and go with `last_supported_delegate_type`.

rseanhall commented:
> Error code when trying to load SCD component

The reason I thought it was acceptable is because there is no way for the native host to distinguish this from the case where the config file is actually invalid. So I think their code was brittle to begin with. But I understand if we can't take this change.

I'm also making `hostfxr_get_runtime_delegate` return the new `HostApiUnsupportedScenario` error code when called with an app context requesting an unsupported delegate. Is that OK?

vitek-karas commented:
> I'm also making hostfxr_get_runtime_delegate return the new HostApiUnsupportedScenario error code when called with an app context requesting an unsupported delegate. Is that OK?

For some reason I think it is more OK than the other :shrug:. This one is a clear programmer error, so it's unlikely there will be fallback code paths which react to the error code to do something meaningful. The first one is data driven - depends on the `.runtimeconfig.json` so it's frequently out of control for the code itself and thus the code might have a fallback code path.

So yes - I'm OK taking the change on `hostfxr_get_runtime_delegate`.

vitek-karas commented on src/installer/corehost/cli/fxr/fx_muxer.cpp:
```diff
@@ -870,8 +870,26 @@ int fx_muxer_t::run_app(host_context_t *context)
+    if (type > coreclr_delegate_type::load_assembly_and_get_function_pointer
+        && (size_t)type >= context->hostpolicy_context_contract.unknown_delegate_type)
+    {
+        return StatusCode::HostApiUnsupportedVersion;
```
I think we should print out an error message in this case. To make it easier to diagnose what happened (the version of `hostpolicy` in use is not always apparent when writing the calling code for this).

vitek-karas commented on src/installer/corehost/cli/hostpolicy/hostpolicy.cpp:
```diff
@@ -744,6 +744,7 @@ SHARED_API int HOSTPOLICY_CALLTYPE corehost_initialize(const corehost_initialize
             rc = StatusCode::Success_DifferentRuntimeProperties;
     }
 
+    auto version_lo = context_contract->version;
```
I think this is a problem. Shipped versions of `hostfxr` (3.0) would not initialize this value on input. So `hostpolicy` must not rely on the value being correct (it's effectively random data).

We will need to find a different way for hostfxr to pass along the version information. The only one I can think of is to add a new bit to the `initialization_options_t` - something like `versioned_contract` or something similar which would tell hostpolicy that it's OK to look at the version field - in fact in that case I think `hostfxr` should initialize all the fields, so that in the future we can use them to communicate intent both ways if needed.

This is unfortunate - we should have set the version in 3.0 already, the contract as is, is effectively broken because the allocated size comes from `hostfxr`, but it's filled with `hostpolicy` without telling it how big the buffer is. I'm starting to question if we should introduce a new API just to fix this.

rseanhall replied:
I initially considered passing the size of the structure into `initialization_options_t` but I thought this would be easier. I'm not used to working in code that doesn't *always* initialize variables, so I forgot you couldn't rely on it being all zeros. I'm going to hold off on making any more changes until you guys decide how you want to handle this.

vitek-karas replied:
@elinor-fung @AaronRobinsonMSFT @swaroop-sridhar any ideas?

The short story: in 3.0 hostfxr calls hostpolicy's initialize with a pointer to a structure which hostpolicy fills in. The structure has a version field, but it's only initialized on the hostpolicy side. So if we decide to add more fields to the struct in 5.0, there's no way to pass the size of the allocated struct from hostfxr to hostpolicy - since hostpolicy can't trust the version field on input (as 3.0 hostfxr didn't initialize it).

Options:
* Introduce a new initialize API on hostpolicy which should probably have one more parameter - the size of the struct. Just like any other out buffer param should have.
* Add a new bit to the initialize options which would tell hostpolicy that newer hostfxr is calling and that the version field in the struct is initialized and thus hostpolicy can trust it (basically a version flag).
* ???

We could also do this somehow via the global initialization of hostpolicy (load), but I don't like that very much.

AaronRobinsonMSFT replied:
I am unclear why this is an issue. That context is intentionally an out parameter that informs hostfxr of the contract that hostpolicy has. This API isn't a conversation it is a question - "tell me what you support". The caller needs to check the version and can then use the struct however it is filled out.

vitek-karas replied:
The problem is the size of the out parameter. The memory is caller allocated (the caller passes a pointer to a piece of memory) but the method doesn't take the size of the allocated memory. Right now it works because both caller and callee agree on the structure. If we want to version that structure (it has a version field, so it was designed for this), we also need to somehow pass how much memory is available, so that the callee can write more data into it, if there's space. Currently we don't do that, so the callee must always write exactly the same number of bytes it always did.

AaronRobinsonMSFT replied:
Spoke offline and finally understand the issue. I would suggest providing a new flag that is something like `version_set` or something like that.

vitek-karas replied:
Offline discussion with @AaronRobinsonMSFT:
* It is possible to have older `hostfxr` talking to a newer `hostpolicy` (private install of older runtime - running dotnet.exe from it - DOTNET_MULTILEVEL_LOOKUP=1 which is the default - trying to run an app - the older hostfxr will now also search global install location for frameworks - if the app allows it it may roll-forward to a newer framework from global install - that has hostpolicy which is newer than the hostfxr).

* Currently hostpolicy doesn't know what version of hostfxr it's talking to (we have version fields on the host_interface_t but they're effectively useless for this purpose, at least the way they're used now).

* Would like: Add a full version of hostfxr into the host_interface_t - could be used by hostpolicy in case we need to quirk something (let's hope not, but seeing this we might). Pass it along to the runtime which could trace it for auditing purposes, or diag purposes and so on (currently there's no easy way to tell the version of hostfxr used to run a given app). This should be done in a separate PR as it's not needed for this change.

* Solve this current problem by adding a bit to the `options` parameter for `corehost_initialize` - something like `initialization_options_t::context_contract_version_set = 0x80000000`. hostpolicy will check that flag and if set it will assume the `corehost_context_contract::version` field is initialized on input (what this PR already does) and will use it to determine if it can fill the struct with the new fields or not. If the bit is not set it will not read from the buffer at all (random data there) and instead assume the "old" version of the structure - basically it can assume `version=sizeof(size_t)*7`.

rseanhall replied:
I can start trying to add that bit to `intialization_options_t` (it's misspelled :( ). (In my original response I didn't look at the code and assumed that this was the global initialization struct, my initial thought was to add something to `host_interface_t` but this should work just as well).

AaronRobinsonMSFT replied:
>intialization_options_t (it's misspelled :( )

You can fix that spelling if you like. It doesn't represent breaking compat change.

vitek-karas replied:
Can you please also fix the spelling - that is perfectly doable as it's only a build-time thing (the hostpolicy header is internal, nobody but hostfxr should be calling hostpolicy directly).

vitek-karas commented on src/installer/corehost/cli/fxr/standalone/hostpolicy_resolver.cpp:
```diff
@@ -201,8 +201,8 @@ int hostpolicy_resolver::load(
 
-        // It's possible to not have corehost_set_error_writer and corehost_initialize. These were
-        // introduced in 3.0, so 2.0 hostpolicy would not have the exports. In this case, we will
+        // It's possible to not have corehost_set_error_writer, corehost_initialize, corehost_main, and corehost_main_with_output_buffer.
```
`corehost_main` has been there since forever... and `corehost_main_with_output` has been added in 2.0 (if I remember correctly).

vitek-karas commented on src/installer/corehost/error_codes.h:
```diff
@@ -51,6 +51,7 @@ enum StatusCode
     HostInvalidState                    = 0x800080a3,
     HostPropertyNotFound                = 0x800080a4,
     CoreHostIncompatibleConfig          = 0x800080a5,
+    HostApiUnsupportedScenario          = 0x800080a6,
```
Can you please add this to https://github.com/dotnet/runtime/blob/master/docs/design/features/host-error-codes.md

AaronRobinsonMSFT commented on src/installer/corehost/cli/fxr/hostfxr.cpp:
```diff
@@ -636,7 +636,11 @@ namespace
+// then all delegate types are supported.
+// If the host_context_handle was initialized using hostfxr_initialize_for_dotnet_command_line,
+// then only the following delegate types are currently supported:
+//     hdt_load_assembly_and_get_function_pointer
```
This is something for the header file for documentation. I wouldn't add public policy decisions in the implementation since we inevitable need to rummage around and update both official documentation and implementation which is confusing. This support isn't hidden so let the code and official doc do the talking.

rseanhall replied:
I agree that this belongs in the header. However, all of this currently lives here. Should I delete all of the comments on public methods in this file and move them to hostfxr.h?

AaronRobinsonMSFT replied:
Oh boo. I didn't realize all comments for APIs were here. You are free to leave this if following the existing convention. Thanks for the attention to detail.

vitek-karas commented on src/installer/corehost/cli/corehost_context_contract.h:
```diff
@@ -9,11 +9,12 @@
-    get_contract = 0x2,          // Get the contract for the initialized hostpolicy
+    wait_for_initialized = 0x1,                // Wait until initialization through a different request is completed
+    get_contract = 0x2,                        // Get the contract for the initialized hostpolicy
+    context_contract_version_set = 0x80000000, // The version field has been set in the corehost_context_contract
```
```suggestion
    context_contract_version_set = 0x80000000, // The version field has been set in the corehost_context_contract on input indicating the maximum size of the buffer which can be filled
```

vitek-karas commented on src/installer/corehost/cli/hostpolicy/hostpolicy.cpp:
```diff
@@ -744,7 +745,7 @@ SHARED_API int HOSTPOLICY_CALLTYPE corehost_initialize(const corehost_initialize
     }
 
-    auto version_lo = context_contract->version;
+    auto version_lo = version_set ? context_contract->version : 7 * sizeof(size_t);
```
Please add a comment what the `7 * sizeof(site_t)` means - otherwise it looks like magic.

vitek-karas commented on src/installer/corehost/cli/hostpolicy/hostpolicy.cpp:
```diff
@@ -752,6 +754,14 @@ SHARED_API int HOSTPOLICY_CALLTYPE corehost_initialize(const corehost_initialize
     context_contract->run_app = run_app;
     context_contract->get_runtime_delegate = get_delegate;
 
+    //An old hostfxr may not have provided enough space for these fields.
+    //The version_lo (sizeof) the old hostfxr saw at build time will be
+    //smaller and we should not attempt to write the fields in that case.
```
Nit: We tend to put space after `// `

vitek-karas reviewed (approved):
Looks good to me - thanks a lot for the change!
Please wait a day or two for @AaronRobinsonMSFT  and @elinor-fung  to have a change to take a look as well.

elinor-fung commented on src/installer/corehost/cli/fxr/fx_muxer.cpp:
```diff
@@ -870,8 +870,27 @@ int fx_muxer_t::run_app(host_context_t *context)
+        break;
+    }
+
+    if (type > coreclr_delegate_type::load_assembly_and_get_function_pointer
```
Nit: could you add a short comment (similar to what you have in hostpolicy.cpp) pointing out how we can't just always use the `last_known_delegate_type` and the significance of `load_assembly_and_get_function_pointer` as the 'original' last value.

elinor-fung reviewed (approved):
Thank you!

rseanhall commented:
I added the comment about `load_assembly_and_get_function_pointer` and squashed this into two commits. If you want to see all the individual commits, they're at https://github.com/rseanhall/runtime/tree/35465-app-get-runtime-delegate-3.

vitek-karas commented:
@rseanhall thanks a lot for the changes!
